src/librustdoc/doctree.rs

```diff
@@ -29,6 +60,37 @@ impl Module<'hir> {
             foreigns: Vec::new(),
             macros: Vec::new(),
             is_crate: false,
+            from_glob: false,
+        }
+    }
+
+    pub(crate) fn push_item(&mut self, new_item: Item<'hir>) {
+        for item_iter in self.items.iter_mut() {
+            if item_iter.name() == new_item.name() {
+                if item_iter.from_glob {
+                    debug!("push_item: {:?} shadowed by {:?}", *item_iter, new_item);
+                    *item_iter = new_item;
+                    return;
+                } else if new_item.from_glob {
+                    return;
+                }
+            }
+        }
+        self.items.push(new_item);
+    }
+
+    pub(crate) fn push_mod(&mut self, new_item: Module<'hir>) {
+        for item_iter in self.mods.iter_mut() {
+            if item_iter.name == new_item.name {
```

Comment:
```suggestion
        if let Some(shadowed_mod) = self.mods.iter_mut().find(|mod_| mod_.name == new_item.name) {
```

Reply:
Sorry - this won't compile on its own, you'll have to also update the rest of the loop.